schemas/activity-stream/impression-stats/impression-stats.1.schema.json

```diff
@@ -1,5 +1,13 @@
 {
   "$schema": "http://json-schema.org/draft-04/schema#",
+  "mozPipelineMetadata": {
+    "bq_dataset_family": "activity_stream",
```

Comment:
I agree with you that the ideal state here is the m-p-s and m-s-g are fully responsible for normalization, and that the ops logic and pipeline logic no longer have to do any normalization. It's worth spending some time figuring out if that can be the case.

From the pipeline perspective, the decoder only cares about JSON schemas and the BQ sinks only care about BQ schemas. These could be two separate artifacts where the directory structure for JSON schemas is `namespace/doctype` and the directory structure for BQ schemas is `bq_dataset_family/bq_table`.

The ops logic, however, needs to know about both JSON schemas and BQ schemas, and the relationship between them. Currently, we handle that by keeping JSON and BQ schemas in the same directory with the `s/schema.json/bq/` relationship in naming.

What if we take a middle ground between the above two approaches. We keep the current structure of the `generated-schemas` branch of m-p-s, but we add a top-level `bq` directory. During a transition period, we duplicate each `bq` schema, writing it into `schemas`, but also writing it into a path under `bq` that is determined by the `bq_*` metadata fields. So we have `schemas/telemetry/xfocsp-error-report/xfocsp-error-report.1.bq` and also `bq/xfocsp_error_report/xfocsp_error_report_v1.bq`.

We could transition tooling piece-by-piece to use the BQ schemas under `bq` and then eventually drop BQ schemas from from `schemas` or convert them to symlinks purely as an affordance for humans to understand the routing. We could also provide symlinks under `bq` to the relevant JSON schemas for human readability.

This assumes a 1:1 mapping between doctype and table. `xfocsp-error-report` currently breaks that, but I believe we can fix that up and enforce it as a requirement going forward.

> While gcp labels are constrained to under 64 characters, we could probably come up with labeling schemes that would make it easier to work backward from a BQ table to files in generated-schemas

I think it would be great to apply labels `document_namespace` (with value like `activity-stream`), `document_type` (with value like `impression-stats`), and `document_version` (with value like `4`) which could be used to construct the json schema path in m-p-s. We could also inject the entire URL into the table's `description` similar to how we inject m-p-s URLs into dataset descriptions.

Reply:
Ok, so it sounds like you're in favor of "change MPS filename conventions (breaking change)", but deploying it in a piecewise (non-breaking) fashion. I think the approach you outline is reasonable, but I would like some opinions about this structure from other people involved with MSG infrastructure such as @acmiyaguchi and @fbertsch. Generally, I'm in favor of keeping a single artifact (with two directory prefixes e.g. `bq` and `schemas`) i.e. no ops changes to the way `generated-schemas` is turned into a single versioned tarball for ops and pipeline consumers to use (the sink and decoder use the same artifact but read different paths).

From the ops pipeline perspective, only having to know about json schemas and using metadata to find the associated bq schema file path is pretty close to the ideal state I'd like to get to. The ops pipeline may still care about non-normalized values derived from json schemas directory structure for descriptions and labels, but I think it might be useful to eventually move away from any dependence on directory structure in favor of metadata. However, I also don't like the idea of un-normalizing or having both normalized and non-normalized values in metadata, but between these two options, I would prefer the latter.

> This assumes a 1:1 mapping between doctype and table. xfocsp-error-report currently breaks that, but I believe we can fix that up and enforce it as a requirement going forward.

I agree this is not a pressing concern and that it should be true in the future.

Reply:
I'm not against having a separate directory with the bigquery schemas, assuming a single tarball with the artifacts. 

> We could transition tooling piece-by-piece to use the BQ schemas under bq and then eventually drop BQ schemas from from schemas or convert them to symlinks purely as an affordance for humans to understand the routing. We could also provide symlinks under bq to the relevant JSON schemas for human readability.

How would we map backwards from the normalized name to the original naming? There would need to be an extra step as the schemas are being generated?

Reply:
> How would we map backwards from the normalized name to the original naming? There would need to be an extra step as the schemas are being generated?

If we want to provide this mapping, we would do so as the schemas are being generated. As we iterate through the JSON schemas and write out BQ schemas under `bq/`, we could add a symlink or a text file that contains the path to the relevant JSON schema (with non-normalized naming).

Reply:
> How would we map backwards from the normalized name to the original naming? There would need to be an extra step as the schemas are being generated?

Yes the intent here is to avoid having to unnormalize at all, and what klukas describes is how we'd achieve that. We only want to normalize once.

Reply:
The above approach seems best to me. Approaching this piece-meal will make the implementation much easier 👍